### ISSUE TITLE
Type instability with `Base.promote_type` and Static.jl types for Julia v1.10-beta2 and `--check-bounds=no`

### ISSUE THREAD

**Issue by sloede** [JuliaLang/julia#50985]
When using Julia v1.10-beta2 with `--check-bounds=no`, using Static.jl types together with native types causes a type instability. In fact, it looks like `Base.promote_type` is not able to infer a concrete type at all anymore and just returns `Any`.

**MWE (with `julia-1.10-beta2 --check=bounds=no`):**
```julia
julia> using Pkg

julia> Pkg.activate(; temp=true, io=devnull)

julia> Pkg.add(name="Static", version="0.8.8", io=devnull)

julia> using Static

julia> i = 1; s = static(2);

julia> @code_warntype Base.promote_type(typeof(i), typeof(s))
MethodInstance for promote_type(::Type{Int64}, ::Type{StaticInt{2}})
  from promote_type(::Type{T}, ::Type{S}) where {T, S} @ Base promotion.jl:306
Static Parameters
  T = Int64
  S = StaticInt{2}
Arguments
  #self#::Core.Const(promote_type)
  _::Core.Const(Int64)
  _::Core.Const(StaticInt{2})
Body::Any
1 ─      nothing
│   %2 = $(Expr(:static_parameter, 1))::Core.Const(Int64)
│   %3 = $(Expr(:static_parameter, 2))::Core.Const(StaticInt{2})
│   %4 = Base.promote_rule($(Expr(:static_parameter, 1)), $(Expr(:static_parameter, 2)))::Core.Const(Union{})
│   %5 = Base.promote_rule($(Expr(:static_parameter, 2)), $(Expr(:static_parameter, 1)))::Core.Const(Union{})
│   %6 = Base.promote_result(%2, %3, %4, %5)::Any
└──      return %6
```

The error does _not_ occur if I use Julia v1.10 without `--check-bounds=no` or when using Julia v1.9 (with or without `--check-bounds=no`). On those cases, we get the following output for `@code_warntype`:
```julia
julia> @code_warntype Base.promote_type(typeof(i), typeof(s))
MethodInstance for promote_type(::Type{Int64}, ::Type{StaticInt{2}})
  from promote_type(::Type{T}, ::Type{S}) where {T, S} @ Base promotion.jl:306
Static Parameters
  T = Int64
  S = StaticInt{2}
Arguments
  #self#::Core.Const(promote_type)
  _::Core.Const(Int64)
  _::Core.Const(StaticInt{2})
Body::Type{Number}
1 ─      nothing
│   %2 = $(Expr(:static_parameter, 1))::Core.Const(Int64)
│   %3 = $(Expr(:static_parameter, 2))::Core.Const(StaticInt{2})
│   %4 = Base.promote_rule($(Expr(:static_parameter, 1)), $(Expr(:static_parameter, 2)))::Core.Const(Union{})
│   %5 = Base.promote_rule($(Expr(:static_parameter, 2)), $(Expr(:static_parameter, 1)))::Core.Const(Union{})
│   %6 = Base.promote_result(%2, %3, %4, %5)::Core.Const(Number)
└──      return %6
```

The issue first appeared when trying to use `PtrArray` from StrideArrays.jl (see also https://github.com/JuliaSIMD/StrideArrays.jl/issues/78). There, the failure to determine a usable promotion type propagates into a regular array access returning values of type `Any`, making the code unusably slow.

Thanks to @ranocha for creating a more minimum MWE. Maybe @vchuravy has an idea how to fix this (since he resolved a similar issue for v1.9 and StaticArrays.jl)?

cc @chriselrod

**Comment by brenhinkeller** [comment]
I take it this is distinct from (though possibly indirectly related to) the issue discussed in https://github.com/JuliaLang/julia/pull/50239?

**Comment by vchuravy** [comment]
On 1.9 with `--check-bounds=no`
```
 • %2 = < pure > typejoin(::Type{…},::Type{…})::Core.Const(Number) (+c,+e,+n,+t,+s,+m,+i)
```

On 1.10 with `--check-bounds=no`
```
 • %1 = invoke typejoin(::Type{Int64},::Type{StaticInt{2}})::Any (+c,+e,!n,+t,+s,+m,+i)
```

This might be #48684 cc: @aviatesk

**Comment by vchuravy** [comment]
Kinda, https://github.com/JuliaLang/julia/pull/48684 replaced special handling of the almost intrinsics. So it's more than just a removal of pure.

**Comment by vchuravy** [comment]
Putting this on the milestone since we will need to fix this in either `1.10.0` or `1.10.x`

**Comment by KristofferC** [comment]
Talking with @Keno, it seems that this cannot really be fixed in 1.10 and has to wait for 1.11, see e.g. https://github.com/JuliaLang/julia/pull/50641. So I will take this off the milestone.

**Comment by sloede** [comment]
@mbauman Thanks for taking an interest! Since this has likely come up in yet another (possibly related) scenario in JuliaSIMD/StrideArrays.jl#93, do you see any chance that someone could take a stab at the underlying issue, e.g., by following your direction?

**Comment by StefanKarpinski** [comment]
> Can we promise that typejoin is no throw?

Seems reasonable to me. I can't think of any reason for it to throw. If can always just return `Any`.

**Comment by vtjnash** [comment]
It appear that it throws if the inputs are nonsensical

**Comment by sloede** [comment]
I assume `promote_type` will throw (and a quick check seems to confirm) if supplied with non-type arguments:
```julia-repl
julia> typejoin(1, 2)
ERROR: TypeError: in <:, expected Type, got a value of type Int64
Stacktrace:
 [1] typejoin(a::Any, b::Any)
   @ Base ./promotion.jl:31
 [2] top-level scope
   @ REPL[9]:1
```
Is there a way to catch this case early and make the rest of `typejoin` `nothrow`? I am _really_ not an expert on the effects system, thus this may very well be an overly naive question...

**Comment by vtjnash** [comment]
I'd thought you added it to the milestone with an offer to fix it. If not, seems like we should just close this.

**Comment by mbauman** [comment]
No, Valentin added it to the milestone and I spitballed a fix. It appeared as though your answer was that it was an invalid fix, but I'm still not sure.

It does seem to have caused considerable issues throughout the ecosystem (see all the trackbacks above), so I think it'd be great to fix it.  It's a fairly fundamental — but limited! — bit of machinery, and it would seem that there _could be_ a relatively simple fix here.  But I'd need less cryptic guidance in order to actually make a valid change.
